lib/elixir/src/elixir_with.erl

```diff
@@ -39,12 +47,25 @@ expand({'<-', Meta, [Left, Right]}, E) ->
 expand(X, E) ->
   elixir_exp:expand(X, E).
 
+expand_elses(KV, E) when is_list(KV) ->
+  {[{do, EClauses}], EC} = elixir_exp_clauses:'case'([], [{do, KV}], E),
+  {[{else, EClauses}], EC};
+expand_elses(nil, E) ->
+  {[], E}.
+
 %% Translation
 
-translate(_Meta, Args, S) ->
-  {Parts, [{do, Expr} | _Opts]} = elixir_utils:split_last(Args),
-  Cases = build_cases(Parts, Expr),
-  elixir_translator:translate(Cases, S).
+translate(Meta, Args, S) ->
+  {Parts, ExprList} = elixir_utils:split_last(Args),
+  CaseExpr =
+    case ExprList of
+      [{do, Expr}, {else, ElseExpr} | _Opts] ->
+        Cases = build_cases(Parts, {?ref(Meta), Expr}),
```

Comment:
Also we could change name `reference` –> `make_ref` to show what's the actual call inside.